src/mlpack/core/data/dictionary_encoding_impl.hpp

```diff
@@ -0,0 +1,167 @@
+    DicitonaryEncoding &old_obj)
+{
+  if (this != &old_obj)
+  {
+    this->mappings.clear();
+    this->stringq.clear();
+    for (auto it = old_obj.mappings.begin(); it != old_obj.mappings.end(); it++)
+    {
+      this->stringq.push_back(std::string(it->first));
+      this->mappings[stringq.back()] = it->second;
+    }
+  }
+  else
+  {
+    this->mappings = old_obj.mappings;
+    this->stringq = old_obj.stringq;
+  }
+}
+
+template<typename TokenizerType>
+void DicitonaryEncoding::CreateMap(std::string& strings,
+    TokenizerType tokenizer)
+{
+  boost::string_view strView(strings);
+  boost::string_view token;
+  token = tokenizer(strView);
+  std::size_t curLabels = mappings.size() + 1;
+  while (!token.empty())
+  {
+    if (mappings.find(token) == mappings.end())
+    {
+        stringq.push_back(std::string(token));
+        mappings[stringq.back()] = curLabels++;
+    }
+    token = tokenizer(strView);
+  }
+}
+
+template<typename MatType, typename TokenizerType>
+void DicitonaryEncoding::Encode(const std::vector<std::string>& strings,
+                MatType& output, TokenizerType tokenizer)
+{
+  boost::string_view strView;
+  boost::string_view token;
+  std::vector< std::vector<boost::string_view> > dataset;
+  size_t colSize = 0;
+  for (size_t i = 0; i < strings.size(); i++)
+  {
+    strView = strings[i];
+    token = tokenizer(strView);
+    dataset.push_back(std::vector<boost::string_view>() );
+    while (!token.empty())
+    {
+      dataset[i].push_back(token);
+      token = tokenizer(strView);
+    }
+    colSize = std::max(colSize, dataset[i].size());
+  }
+  size_t curLabel = mappings.size() + 1;
+  output.zeros(dataset.size(), colSize);
+  for (size_t i = 0; i < dataset.size(); ++i)
+  {
+    for (size_t j = 0; j < dataset[i].size(); ++j)
+    {
+      if (mappings.find(dataset[i][j]) == mappings.end())
+      {
+        stringq.push_back(std::string(dataset[i][j]));
+        mappings[stringq.back()] = curLabel++;
+      }
+      output.at(i, j) = mappings.at(dataset[i][j]);
+    }
+  }
+}
+
+template<typename TokenizerType>
+void DicitonaryEncoding::Encode(const std::vector<std::string>& strings,
+            std::vector<std::vector<size_t> >& output, TokenizerType tokenizer)
+{
+  boost::string_view strView;
+  boost::string_view token;
+  size_t curLabel = mappings.size() + 1;
+  for (size_t i = 0; i < strings.size(); ++i)
+  {
+    output.push_back(std::vector<size_t>() );
+    strView = strings[i];
+    token = tokenizer(strView);
+    while (!token.empty())
+    {
+      if (mappings.count(token) == 0)
+      {
+        stringq.push_back(std::string(token));
+        mappings[stringq.back()] = curLabel++;
+      }
+    output[i].push_back(mappings.at(token));
+    token = tokenizer(strView);
+    }
+  }
+}
+
+template<typename Archive>
+void DicitonaryEncoding::serialize(Archive& ar, const unsigned int
+    /* version */)
+{
+  std::unordered_map<std::string, size_t> temp;
```

Comment:
Actually, you can avoid extra memory allocation (I mean the `temp` variable). You should serialize a string and its label sequentially.

Reply:
That was a good idea.